src/backend/x11/mod.rs

```diff
@@ -0,0 +1,921 @@
+            .map(|(_, depth)| depth)
+            .ok_or(CreateWindowError::NoDepth)?;
+
+        // Next find a visual using the supported depth
+        let visual_id = depth
+            .visuals
+            .iter()
+            .find(|visual| visual.class == VisualClass::TRUE_COLOR)
+            .ok_or(CreateWindowError::NoVisual)?
+            .visual_id;
+
+        let format = match depth.depth {
+            24 => DrmFourcc::Xrgb8888,
+            32 => DrmFourcc::Argb8888,
+            _ => unreachable!(),
+        };
+
+        // Make a colormap
+        let colormap = connection.generate_id()?;
+        connection.create_colormap(ColormapAlloc::NONE, colormap, screen.root, visual_id)?;
+
+        let atoms = Atoms::new(&*connection)?.reply()?;
+
+        let window = Arc::new(WindowInner::new(
+            Arc::downgrade(&connection),
+            screen,
+            properties,
+            atoms,
+            depth.clone(),
+            visual_id,
+            colormap,
+        )?);
+
+        let source = X11Source::new(
+            connection.clone(),
+            window.id,
+            atoms._SMITHAY_X11_BACKEND_CLOSE,
+            logger.clone(),
+        );
+
+        info!(logger, "Window created");
+
+        let (resize_send, resize_recv) = mpsc::channel();
+
+        let backend = X11Backend {
+            log: logger,
+            source,
+            connection,
+            window,
+            key_counter: Arc::new(AtomicU32::new(0)),
+            depth,
+            visual_id,
+            screen_number,
+            resize: resize_send,
+        };
+
+        let surface = X11Surface::new(&backend, format, resize_recv)?;
+
+        Ok((backend, surface))
+    }
+
+    /// Returns the default screen number of the X server.
+    pub fn screen(&self) -> usize {
+        self.screen_number
+    }
+
+    /// Returns the underlying connection to the X server.
+    pub fn connection(&self) -> &RustConnection {
+        &*self.connection
+    }
+
+    /// Returns a handle to the X11 window this input backend handles inputs for.
+    pub fn window(&self) -> Window {
+        self.window.clone().into()
+    }
+}
+
+/// An X11 surface which uses GBM to allocate and present buffers.
+#[derive(Debug)]
+pub struct X11Surface {
+    connection: Weak<RustConnection>,
+    window: Window,
+    resize: Receiver<Size<u16, Logical>>,
+    device: gbm::Device<RawFd>,
+    format: DrmFourcc,
+    width: u16,
+    height: u16,
+    current: Dmabuf,
+    next: Dmabuf,
+}
+
+impl X11Surface {
+    fn new(
+        backend: &X11Backend,
+        format: DrmFourcc,
+        resize: Receiver<Size<u16, Logical>>,
+    ) -> Result<X11Surface, X11Error> {
+        let connection = &backend.connection;
+        let window = backend.window();
+
+        // Determine which drm-device the Display is using.
+        let screen = &connection.setup().roots[backend.screen()];
+        let dri3 = connection.dri3_open(screen.root, 0)?.reply()?;
```

Comment:
Could you add a comment explaining the `0`? https://github.com/freedesktop/xorg-dri3proto/blob/master/dri3proto.txt says this must be a "RandR provider". I don't think I ever heard that concept before.

Seems like you just pick the first one here and servers usually just have one:
```
$ xrandr --listproviders
Providers: number : 1
Provider 0: id: 0x47 cap: 0xf, Source Output, Sink Output, Source Offload, Sink Offload crtcs: 3 outputs: 5 associated providers: 0 name:modesetting
```
Hm. But here, the provider actually has `id: 0x47`...

Ohh, you are actually using a special case in the X11 server that is not documented in the spec:

https://gitlab.freedesktop.org/xorg/xserver/-/blob/6c1e6429bcc4447601d9327c84513421226c45b8/dri3/dri3_request.c#L138

Could you please use `x11rb::NONE` instead of the `0` then? Numerically, that is the same thing, but it makes it a tiny bit clearer that this requests expects an XID.